src/sensors/lorreconstruction.rs

```diff
@@ -0,0 +1,404 @@
+        Ok(LorBatch::new(lors_from(&events, |evs| lor_from_sensor_positions(evs, &xyzs, threshold, min_sensors, min_charge)), events.len()))
+    })
+}
+
+// Function to correct for DOI
+// Takes into account the physical range for the setup.
+// TODO This is not general enough for phi rms
+fn calculate_interaction_position(m: Ratio, c: Length, rmin: Length, rmax: Length) -> impl Fn(Length, Length, Length) -> (Length, Length) {
+    move |rms: Length, x: Length, y: Length| -> (Length, Length) {
+        let mut r = m * rms + c;// Bias correction?
+        if r < rmin { r = rmin; }
+        else if r > rmax { r = rmax; }
+        let phi = y.atan2(x);
+        (r * phi.cos(), r * phi.sin())
+    }
+}
+
+// Calculate the azimuthal angle for each hit and
+// adjust so continuous (make option?)
+fn azimuthal_angle(hits: &[MCQT], xyzs: &SensorMap) -> Option<Vec<Angle>> {
+    if hits.is_empty() { return None }
+    let nhit = hits.len();
+    let mut phis = Vec::with_capacity(nhit);
+    for &MCQT { sensor_id, .. } in hits {
+        let &(x, y, _) = xyzs.get(&sensor_id)?;
+        phis.push(y.atan2(x));
+    }
+    if phis.iter().any(|&phi| phi < turn(-0.25)) && phis.iter().any(|&phi| phi > turn(0.0)) {
+        // There is a discontinuity in this case (sensors in second and third quadrants) so mean or std calculations would be wrong.
+        for phi in phis.iter_mut() {
+            if *phi < turn(-0.25) {
+                *phi += turn(1.0);
+            }
+        }
+    }
+    Some(phis)
+}
+
+// Weighted mean and variance.
+// Might be better as a trait?
+pub fn charge_weighted_mean<T>(data: &[T], weights: &[Ratio]) -> Option<T>
+where
+    T: Clone + std::iter::Sum + Mul<Ratio> + Div<Ratio>,
+    <T as Mul<Ratio>>::Output: Into<T>,
+    <T as Div<Ratio>>::Output: Into<T>
+{
+    if data.is_empty() { return None }
+    assert_eq!(data.len(), weights.len());
+    let sum_weights: Ratio = weights.iter().cloned().sum();
+    let weighted_sum: T = data.iter().cloned()
+        .zip(weights.iter().cloned())
+        .map(|(d, w)| (d * w).into())
+        .sum();
+    Some((weighted_sum / sum_weights).into())
+}
+
+pub fn charge_weighted_std<T>(data: &[T], weights: &[Ratio], mean: Option<T>) -> Option<T>
+where
+    T: Clone + Real + std::iter::Sum + Sub<Output = T> + Mul<Output = T> + Mul<Ratio> + Div<Ratio>,
+    <T as Mul<Ratio>>::Output: Into<T>,
+    <T as Div<Ratio>>::Output: Into<T>
+{
+    if data.is_empty() { return None }
+    assert_eq!(data.len(), weights.len());
+    let mval = match mean {
+        Some(mval) => mval,
+        None => charge_weighted_mean(data, weights)?
+    };
+    let adj_wsum: Ratio = weights.iter().cloned().sum::<Ratio>() - ratio(1.0);
+    let var: T = data.iter().cloned()
+        .zip(weights.iter().cloned())
+        .map(|(d, w)| {
+            let sq: T = (d.clone() - mval.clone()) * (d - mval.clone());
+            (sq * w).into()
+        })
+        .sum();
+    Some((var / adj_wsum).into().sqrt())
+}
+
```

Comment:
I'm starting to think that this function can't be completely generic due to the `Sqrt` implementation. `Angle` implements it but `Length` doesn't and while `Length * Length` gives `Area`, the same is not true for `Angle` as you'd expect.

I tried having a separate generic type for the `var` variable but then I remembered that `Real` (where `Sqrt` is) forces the output to `Self`.

Am I missing something or do you think I should implement separate functions/a trait?

Reply:
`uom` only implements `sqrt` for quantities with even powers of dimensions. (As mentioned elsewhere, I'm not in a position to look at further details, right now.)

Reply:
You can see the extremely long-winded signature of `Quantity::sqrt` [here](https://docs.rs/uom/latest/uom/si/struct.Quantity.html#method.sqrt). It should be possible to make a generic version of the function by using this sort of consrtaint somewhere. It would be an edifying and character-bulding experience; whether it would be worthwhile, in the current context, remains to be seen.

You can see that I [avoided the issue](https://github.com/jacg/petalo/blob/master/src/fom.rs#L242-L250) in the FOM calculations. The trick would be to unwrap `f32`s from `uom` quantities, pass them to the uomless function, and wrap the result back up in an `uom` quantity straght afterwards. (This is a bit like using `unsafe` in Rust: to be avoided, if possible.)

Reply:
Apologies if I am intruding here, but since you mentioned me a few weeks ago I keep getting the emails :)

Regarding the sqrt problem, I [asked](https://github.com/iliekturtles/uom/issues/302) the maintainer of the repo, and they will eventually add rational powers, but it's not quite there yet.

However, I don't understand why there is a problem here. The variance has dimensions [dimension of input]^2, i.e. even powers of whatever dimensions you are using, so applying the sqrt to that quantity should not be a problem. Am I missing something?

Reply:
I was trying to write a generic function that could calculate both the standard deviation for `Angle` and `Length` types since we had those options in the code I'm translating for the parameterisation of the interaction R calculation. I was hitting against a wall for the trait bounds necessary for the function.

Reply:
Basically the function as is doesn't work (`Length` doesn't implement sqrt), and `Real` is the wrong bound I think anyway... it's a bit of a mess at the moment, sorry.

Reply:
I understand, but it shouldn't matter that `Length` does not implement sqrt, because sqrt is applied on something that is `Length`^2. I'm not sure if this becomes `Area` automatically; I guess not because `Area` does implement sqrt. I think `uom` should allow  `sqrt(x*x)` for all `x` so maybe this is an issue to be raised.

Reply:
Yeah, but you have to define a type for the variance on which you impose certain trait bounds so that you can use functions like `sqrt`. My most recent attempt used a second generic type to be the type of the variance but I can't understand how to impose the `sqrt` trait bound. If I use `Real` it's incorrect because the `Real::Sqrt` implementation imposes a return of type `Self` whereas in `uom` you'd get another type (`Area` -> `Length`). I tried some stuff from the link that @jacg sent (I had found it myself yesterday) but to no avail. The documentation is not really aimed at someone of my level of experience with rust and uom, I fear.

Reply:
> Am I correct in thinking that it works for hard-wired Length, and that the problems only start to appear when you try to make it generic? If so, would you like me to try to implement the generic version (I can dedicate some time to it, and it might be edifying for me)?

I have a version hardwired for `Length` and another for `Angle`, `std_length` and `std_angle` respectively. The difficulty is getting the generic version to work.

Reply:
Here are some demonstrations of how to write `uom`-generic functions that use `sqrt()`

+ https://github.com/jacg/petalo/blob/e71a5235a4b2cddceda7fa7f163b44b30d57ce76/src/fom.rs#L238-L270
+ https://github.com/jacg/petalo/blob/e71a5235a4b2cddceda7fa7f163b44b30d57ce76/src/fom.rs#L286-L317
+ https://github.com/jacg/petalo/blob/e71a5235a4b2cddceda7fa7f163b44b30d57ce76/src/fom.rs#L341-L352
+ https://github.com/jacg/petalo/blob/e71a5235a4b2cddceda7fa7f163b44b30d57ce76/src/fom.rs#L370-L382

+ They are not generic over the underlying value type, but hard-wired for f32. That's fine for our purposes here.
+ Instead of trying to satisfy all the bounds (I tried that approach a few times, and it was a futile exercise in masochism), it uses the hack of unwrapping the value stored inside the quantity, doing the calculation on plain `f32`s, and wrapping it up again in the quantity.

Reply:
Yeah, the generic versions are implemented.